### PR TITLE
Add to favorite list check in log visit (Fixes #4347)

### LINKED ISSUE #4347: Allow adding favpoint in logging activity
From support mail:
Right now I have to first log the cache and afterwards add it to my favorites on another page of c:geo. It would be better to have the option to add it, right on the logging screen like it is realized on the website (and in c:geo for voting on gcvote). It somehow makes sense, to decide about favoriting the cache while logging it, rather than having to do it manually afterwards.

### PR DESCRIPTION
**Pull request by madankb**
Fix for the issue "Allow adding favpoint in logging activity #4347". Please review and share your thought.

### CODE REVIEW

**Comment by kumy** [comment]
@madankb could you please rebase your code on current master.

**Comment by kumy** [comment]
Please follow [c:geo Coding Conventions](https://github.com/cgeo/cgeo/wiki/Coding-Conventions), and [Java Naming Convention](https://en.wikipedia.org/wiki/Naming_convention_%28programming%29#Java).

You missed some `final` statements, and Variables and Functions names as `lowerCamelCase`.


* In `LogCacheActivity` you may add some more code to `Restore from a previous state`, either from *an activity restore* or a *saved log*.

* I better see code for `getfavoritepoints()` in `GCLoggingManager.onLoadFinished()`

**Comment by SammysHP** [comment]
The log activity loads the log page to get the necessary information. Please get the favorite point count from there.

**Comment by madankb** [comment]
@kumy @SammysHP I will change my code as per your suggestion.

**Comment by madankb** [comment]
@kumy In GC, if there is no favorite points left, add to favorites is hidden but not greyed/disabled. Can we follow the same way here?. If it is need to be greyed out, it will be like "Add to Favorites (0 remaining)" greyed out. Please comment.  

Can you please clarify this comment "Note: on GC-website, Add to your Favorites (XX remaining) may be absent if cache 'is mine'" ?. I take the number of favorite points for premium user through "../my/default.aspx?". Can I proceed to parsing in same way to take favorite points?. Please comment.

**Comment by kumy** [comment]
(I am on the mobile phone).

Ok, let it be hidden for now, we'll modify later if necessary.

Please don t parse "default.aspx" but "seek/log.aspx". Add the code in the function I mentioned "GCLoggingManager.onLoadFinished()". This will save us a network call.

I meant that "seek/log.aspx" don t always display the add to favorite checkbox, for example if you try to add favorite point on a cache that belongs to the "connected user".

I also thought this morning that your current implementation may propose users the "add to favorite" if they log on other connector than GC. (However not verified, it was memory). Parsing from "GCLoggingManager.onLoadFinished()" will probably save you this particular case.

(I ll be off until tonight)

**Comment by SammysHP** [comment]
@madankb As I said: We already parse the log page (to get the necessary session information).

File: main/src/cgeo/geocaching/LogCacheActivity.java [hunk 1]
```diff
@@ -256,6 +261,7 @@ public void onCreate(final Bundle savedInstanceState) {
         }
 
         initializeRatingBar();
+        intializeFavoritecheck(cache);
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/LogCacheActivity.java]: Typo ("i" missing in "initialize").

File: main/src/cgeo/geocaching/LogCacheActivity.java [hunk 2]
```diff
@@ -351,6 +357,19 @@ public void onRatingChanged(final float stars) {
         }
     }
 
+    private void intializeFavoritecheck(Geocache cache) {
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/LogCacheActivity.java]: `final` missing (at many places afterwards also).

File: main/src/cgeo/geocaching/LogCacheActivity.java [hunk 3]
```diff
@@ -351,6 +357,19 @@ public void onRatingChanged(final float stars) {
...
+        if (Settings.isGCPremiumMember() && cache.supportsFavoritePoints() && !cache.isOwner()) {
+            String favoritecount = Settings.getPremFavPoints();
+            if (favoritecount != null && Integer.parseInt(favoritecount) > 0) {
+                favbox.setVisibility(View.VISIBLE);
+                favoritecheck.setChecked(true);
+                favoriteremaining.setVisibility(View.VISIBLE);
+                favoriteremaining.setText(getString(R.string.fav_points_remaining, favoritecount));
+            }
+
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/LogCacheActivity.java]: Extra empty line.

File: main/src/cgeo/geocaching/LogCacheActivity.java [hunk 4]
```diff
@@ -496,6 +516,17 @@ protected StatusCode doInBackgroundInternal(final String[] logTexts) {
                         }
                     }
 
+                    if (favoritecheck.isChecked() && favbox.getVisibility() == View.VISIBLE) {
+                        if (GCParser.addToFavorites(cache)) {
+                            DataStore.saveChangedCache(cache);
+                            GCLogin.getInstance().getfavoritepoints();
+                            Log.d("Successfully added to favorites list");
+                        } else {
+                            showToast("Error while adding the cache to Favorite list");
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/LogCacheActivity.java]: Toasts should only contain localized strings.

File: main/src/cgeo/geocaching/connector/gc/GCLogin.java [hunk 5]
```diff
@@ -282,6 +284,21 @@ public void call() {
         });
     }
 
+    public String getfavoritepoints() {
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/connector/gc/GCLogin.java]: Bad caml casing.

File: main/src/cgeo/geocaching/connector/gc/GCLogin.java [hunk 6]
```diff
@@ -282,6 +284,21 @@ public void call() {
...
+        Log.d("Getting favorite points of premium user");
+        final String result = Network.getResponseData(Network.getRequest("https://www.geocaching.com/my/default.aspx?"));
+        if (result != null) {
+            String count = getNumberString(TextUtils.getMatch(result, GCConstants.PATTERN_SEARCH_PREMIUM_FAVORITE, false, 1, null, true)).trim();
+            Settings.setPremFavPoints(count);
+            return result;
+        }
+        return null;
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/connector/gc/GCLogin.java]: Could be `return result;` here and the one two lines above could be removed.

File: main/src/cgeo/geocaching/connector/gc/GCLogin.java [hunk 7]
```diff
@@ -282,6 +284,21 @@ public void call() {
         });
     }
 
+    public String getfavoritepoints() {
+        Log.d("Getting favorite points of premium user");
+        final String result = Network.getResponseData(Network.getRequest("https://www.geocaching.com/my/default.aspx?"));
+        if (result != null) {
+            String count = getNumberString(TextUtils.getMatch(result, GCConstants.PATTERN_SEARCH_PREMIUM_FAVORITE, false, 1, null, true)).trim();
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/connector/gc/GCLogin.java]: A `count` should be manipulated as number, not as text.

File: main/src/cgeo/geocaching/settings/Settings.java [hunk 8]
```diff
@@ -1196,4 +1196,15 @@ public static String getHomeLocation() {
     public static void setHomeLocation(@NonNull final String homeLocation) {
         putString(R.string.pref_home_location, homeLocation);
     }
+
+    public static void setPremFavPoints(final String FavPoints) {
```
Comment (samueltardieu) [inline main/src/cgeo/geocaching/settings/Settings.java]: Parameter name should start with a lowercase.

**Comment by madankb** [comment]
@kumy I tried with the value ("ctl00$ContentBody$LogBookPanel1$chkAddToFavorites","on") and ("ctl00$ContentBody$LogBookPanel1$chkAddToFavorites","checked") in params field mentioned by you while requesting http://www.geocaching.com/seek/log.aspx?. But still the cache is not added into favorite list. Please share some ideas to find the issue. Appreciate your help.

**Comment by rsudev** [comment]
@madankb Could put the real code you use to try this out somewhere (either a branch in your repo or a gist or similar. That would make it easier to understand and discuss.

**Comment by kumy** [comment]
@madankb I've played a bit with this part.
For me adding (statically) the `"ctl00$ContentBody$LogBookPanel1$chkAddToFavorites", "on"` and tested some post on a cache that IS NOT mine (this is important, as FP cannot be added to personal caches) **works fine**.

During the tests, I've found that submitting a "Write Note" doesn't add a cache to Favorites, this doesn't work on GC website too, even if the checkbox is displayed.

As @rsudev proposed, please share your actual code so we have concrete thing to discuss on.

Here is my (dirty hack to do the tests):
```
$ git diff
diff --git a/main/src/cgeo/geocaching/connector/gc/GCParser.java b/main/src/cgeo/geocaching/connector/gc/GCParser.java
index ad33a93..e613b00 100644
--- a/main/src/cgeo/geocaching/connector/gc/GCParser.java
+++ b/main/src/cgeo/geocaching/connector/gc/GCParser.java
@@ -1079,6 +1079,7 @@ public final class GCParser {
                 "ctl00$ContentBody$LogBookPanel1$uxLogInfo", logInfo,
                 "ctl00$ContentBody$LogBookPanel1$btnSubmitLog", "Submit Log Entry",
                 "ctl00$ContentBody$LogBookPanel1$uxLogCreationSource", "Old",
+                "ctl00$ContentBody$LogBookPanel1$chkAddToFavorites", "on",
                 "ctl00$ContentBody$uxVistOtherListingGC", "");
         GCLogin.putViewstates(params, viewstates);
         if (trackables != null && !trackables.isEmpty()) { //  we have some trackables to proceed

```

I've placed a break point just before sending the log, here they are, so you can compare with yours:
```
0 = {BasicNameValuePair@5659} "__EVENTTARGET="
1 = {BasicNameValuePair@5660} "__EVENTARGUMENT="
2 = {BasicNameValuePair@5661} "__LASTFOCUS="
3 = {BasicNameValuePair@5662} "ctl00$ContentBody$LogBookPanel1$ddLogType=2"
4 = {BasicNameValuePair@5663} "ctl00$ContentBody$LogBookPanel1$uxDateVisited=09/27/2015"
5 = {BasicNameValuePair@5664} "ctl00$ContentBody$LogBookPanel1$uxDateVisited$Month=9"
6 = {BasicNameValuePair@5665} "ctl00$ContentBody$LogBookPanel1$uxDateVisited$Day=27"
7 = {BasicNameValuePair@5666} "ctl00$ContentBody$LogBookPanel1$uxDateVisited$Year=2015"
8 = {BasicNameValuePair@5667} "ctl00$ContentBody$LogBookPanel1$DateTimeLogged=09/27/2015"
9 = {BasicNameValuePair@5668} "ctl00$ContentBody$LogBookPanel1$DateTimeLogged$Month=9"
10 = {BasicNameValuePair@5669} "ctl00$ContentBody$LogBookPanel1$DateTimeLogged$Day=27"
11 = {BasicNameValuePair@5670} "ctl00$ContentBody$LogBookPanel1$DateTimeLogged$Year=2015"
12 = {BasicNameValuePair@5671} "ctl00$ContentBody$LogBookPanel1$LogButton=Submit Log Entry"
13 = {BasicNameValuePair@5672} "ctl00$ContentBody$LogBookPanel1$uxLogInfo=Test"
14 = {BasicNameValuePair@5673} "ctl00$ContentBody$LogBookPanel1$btnSubmitLog=Submit Log Entry"
15 = {BasicNameValuePair@5674} "ctl00$ContentBody$LogBookPanel1$uxLogCreationSource=Old"
16 = {BasicNameValuePair@5675} "ctl00$ContentBody$LogBookPanel1$chkAddToFavorites=on"
17 = {BasicNameValuePair@5676} "ctl00$ContentBody$uxVistOtherListingGC="
18 = {BasicNameValuePair@5677} "__VIEWSTATE=***CUT***"
20 = {BasicNameValuePair@5679} "__VIEWSTATEFIELDCOUNT=2"
21 = {BasicNameValuePair@5680} "ctl00$ContentBody$LogBookPanel1$uxTrackables$hdnSelectedActions="
22 = {BasicNameValuePair@5681} "ctl00$ContentBody$LogBookPanel1$uxTrackables$hdnCurrentFilter="
```

**Comment by madankb** [comment]
@kumy you are right. It is working for "Found it", but not for "Write note". I was trying with write note and thought my code is not working. 
I just found one more thing; GetFavoritepoints fix is put in GCLoggingManager (Called from OnLoadFinished). Added one of the cache to favorite list (2 was remaining before adding) and tried for another cache but still it is showing as 2 remaining (but 1 was remaining).  Because GCLoggingManager is called only when log visit is clicked, shouldn't we calculate the favorite points before that (May be when user is still in cache description page)?. @kumy @rsudev Please share your thought.

**Comment by kumy** [comment]
@madankb IIRC, FP count for `GetFavoritepoints` is checked at application start. But in your case, you should extract remaining FP from the webpage. @rsudev am I right?

**Comment by rsudev** [comment]
I must admit currently I have no idea where and how we handle the remaining fav points of a user. But regardless should those be retrieved from the log page when it is loaded in the LoggingManager and then put into the UI async as the other items from the log page, not up-front before starting the LogActivity.

**Comment by madankb** [comment]
@kumy @rsudev  I am good with getting favorite points part, I need few clarification on restoring from previous state. If the user checks add to favorite and didn't post the log; then add to favorite should be checked when user comes back. Should I store favorite check (as boolean) in database (through DataStore.java in saveLogOffline() and loadLogOffline())? Please confirm. Appreciate your help.

**Comment by SammysHP** [comment]
Currently we do not store such data. It's the same for images or trackables.

**Comment by rsudev** [comment]
I also would say we should address storing all log items for offline as a separate task.

**Comment by madankb** [comment]
I believe we store log type, text and date offline (via saveLog in LogCacheActivity to saveLogOffline in DataStore).
@kumy has requested to implement restore status of add to favorite (checked or not checked); I am not sure how to implement it. 
Can someone please guide me?

**Comment by kumy** [comment]
When I asked that, I refered to saving the FP status while the Activity is destroyed/recreacted. Eg in a screen rotation or starting a new activity, like displaying a trackable item. IIRC this is done in a saving bundle. (onSaveActivity, onRestoreActivity ?)

On another hand, saving the FP status for offline use will be a good thing too. But here we also have some issue there. Actually, pictures, gcvote, trackables item, are not saved in the datastore. @rsudev Shouldn t this specific part be threated in another "global" PR?
-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by madankb** [comment]
@kumy As of now if Add to favorite points is checked or unchecked; same status is maintained after screen rotation. Only favorite points is recalculated in GCLoggingManager, I need to put fix not to recalculate favorite points when screen is rotated, but to take from onSaveInstanceState (of GCLoggingManager)? Please correct me if my understanding is wrong.

**Comment by kumy** [comment]
@madankb this is probably because Android don't claim the memory. There is an option in the developer settings called ["Don't Keep Activities" / "Immediately destroy activities"](http://developer.android.com/tools/debugging/debugging-devtools.html), you could enable it for testing the behaviour. But don't keep it enabled after testing as this could break other apps.
Also, may you share your actual (rebased) code, to see which point you reached :)

**Comment by madankb** [comment]
@kumy please find the .java files changed in https://gist.github.com/madankb. I implemented the changes as mentioned.  Adding the favorite points and displaying the number of caches remaining is done and restore state hasn't implemented. Please verify the code and share your thought. I rebased my code sometime back, will rebase code after your comments.

**Comment by kumy** [comment]
@madankb may you please `git push`. There is no "diff" in your gist, it's not easy to see what you have changed. Thanks

**Comment by madankb** [comment]
@kumy please review the latest commit and share your suggestions.

**Comment by kumy** [comment]
@madankb thanks, I'll try to have a look tonight

**Comment by madankb** [comment]
@kumy I updated the code as you suggested. I am not sure about the onsavedinstance (to restore the FP). Please verify the LogCacheActivity changes ( https://gist.github.com/madankb ) made to restore favorite points.

**Comment by kumy** [comment]
@madankb At first view it seems good. May you push your full code?

**Comment by madankb** [comment]
@kumy Can you please verify now?

**Comment by kumy** [comment]
> I hope supportsAddToFavorites as generic term.

Sounds good to me.

**Comment by Lineflyer** [comment]
@madankb If you would like to also get feedback for the UI implementation, you might want to post some screenshots here so that I can take a look (not able to compile myself as I am only tester).

It might be more efficient to do it at this stage then only after it has been merged.

**Comment by madankb** [comment]
screenshot_2015-10-15-20-35-18

**Comment by madankb** [comment]
@Lineflyer here you go

**Comment by Lineflyer** [comment]
Thanks @madankb 
From UI perspective this looks like I would habe expected it. 
I guess the checkbox will be disabled if no favorite point is remaining?
Additionally the whole line should be invisible for basic members.

**Comment by madankb** [comment]
@Lineflyer Yes. You are right. @kumy Please review my code.

**Comment by madankb** [comment]
Please review my code. I made the changes as you guys suggested.

**Comment by samueltardieu** [comment]
Looks code to me as far as style is concerned. I haven't tested it (and can't right now) though.

**Comment by madankb** [comment]
@samueltardieu Did you find any issue in testing?.

**Comment by samueltardieu** [comment]
As I said, I can't test right now. If I can, I'll post my results here or merge. But someone will probably beat me in the meantime.

**Comment by kumy** [comment]
@madankb @samueltardieu I'll plan to do it "soon"

**Comment by kumy** [comment]
Code tested and merged.
@madankb thanks for your contribution.

**Pull request merged**